crates/transport_iroh/src/lib.rs

```diff
@@ -373,6 +355,213 @@ impl IrohTransport {
+        // field is updated.
+        //
+        // If the preflight has not been received yet, read the preflight
+        // from the stream. Time out if the preflight isn't received and
+        // return to close the stream.
+        let result = tokio::time::timeout(Duration::from_secs(10), async {
+            let (remote_url, preflight_bytes) = read_preflight_frame_from_stream(&mut recv_stream).await?;
+
+            ctx.set_remote_url(remote_url.clone());
+            ctx.handler()
+                .recv_data(remote_url.clone(), preflight_bytes)
+                .await?;
+            ctx.set_preflight_received();
+            info!(remote = ?remote_url.peer_id(),"preflight received successfully");
+
+            // If the preflight has not been sent yet, it must be the first message
+            // sent back to the remote.
+            if !ctx.preflight_sent() {
+                let maybe_local_url =
+                    local_url.read().expect("poisoned").clone();
+                if let Some(local_url) = maybe_local_url {
+                    let return_preflight =
+                        ctx.handler().peer_connect(remote_url.clone()).await?;
+                    ctx.send_preflight_frame(
+                        local_url.clone(),
+                        return_preflight,
+                    )
+                    .await?;
+                    info!(peer = ?ctx.connection().remote_id(),?local_url, "sent preflight to peer from url");
+                    ctx.set_preflight_sent();
+                }
+            }
+
+            Ok(remote_url)
+        })
+        .await
+        .map_err(|err| {
+            K2Error::other_src("timed out waiting for preflight", err)
+        });
+        match result {
+            Ok(Ok(remote_url)) => {
+                connections
+                    .write()
+                    .expect("poisoned")
+                    .insert(remote_url, ctx.clone());
+            }
+            Ok(Err(err)) | Err(err) => {
+                error!(?err, "failed to receive preflight frame");
+                return Err(err);
+            }
+        }
+    }
+
+    // Keep reading data frames from the stream until it is closed.
+    loop {
+        // Read data frame header
+        let mut header = [0u8; FRAME_HEADER_LEN];
+        if let Err(err) = recv_stream.read_exact(&mut header).await {
+            error!(?err, remote = ?ctx.remote_url(), "error reading data frame header");
+            break;
+        }
+        let (frame_type, data_len) = match decode_frame_header(&header) {
+            Ok(result) => result,
+            Err(err) => {
+                error!(?err, "failed to decode iroh frame header");
+                // If the frame header could not be decoded, continue by reading the next
+                // frame from the stream.
+                continue;
+            }
+        };
+        if frame_type == FrameType::Preflight {
```

Comment:
It is possible now to remove frame types altogether. Frames would just be decoded based on the convention that the first frame of a connection must be the preflight and every subsequent frame must be a data frame.

It's not a big overhead (1 byte bandwidth and little computation) to keep frame types and they are more expressive, but they are not essential any more.

Let me know what you think.